core/src/main/scala/kafka/server/DelayedProduce.scala

```diff
@@ -124,6 +124,19 @@ class DelayedProduce(delayMs: Long,
   }
 }
 
+class SafeDelayedProduce(delayMs: Long,
```

Comment:
I do not have a strong preference in either case, since we need to modify `replica manager` to pass an additional parameter anyways. @hachikuji @junrao Let me know if you have a strong opinion in this (fewer virtual calls?) and I can change.

BTW any other comments besides this one? Just want to cumulate more comments before I rebase again..